### ISSUE TITLE
YOLOv8 class_weights

### ISSUE THREAD

**Issue by ksv87** [ultralytics/ultralytics#8959]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Are you planing do it?

https://github.com/ultralytics/ultralytics/blob/2d513a9e4bf51e961a4199067383d2052f483874/ultralytics/models/yolo/detect/train.py#L84

**Comment by glenn-jocher** [comment]
@ksv87 hello there! 🚀 Thanks for reaching out. As of now, we don't have immediate plans to implement `class_weights` directly within the training script. We generally prioritize features based on community demand, so this may change in the future. 

Meanwhile, should you wish to experiment with class weighting yourself, you can modify the loss calculation in the training script. For example, you could manually apply different weights to different classes in the loss computation based on your requirements. 

Remember, every dataset is unique, and what works best can vary greatly. Therefore, we encourage experimentation and adaptation to your specific needs. 

If you decide to go down that path and need further guidance or if you have any more questions, feel free to ask! Happy coding! 🎉

**Comment by MonteiroAlvesMatheus** [comment]
@glenn-jocher hello! 

i was looking in the ultralytics' files of yolov8 and i can't find the loss calculation to modify it. 

How can i manually apply different weights to different classes using YOLOv8? Can you help me?

**Comment by ksv87** [comment]
> @glenn-jocher hello!
> 
> i was looking in the ultralytics' files of yolov8 and i can't find the loss calculation to modify it.
> 
> How can i manually apply different weights to different classes using YOLOv8? Can you help me?

may be this [ultralytics/utils/loss.py](https://github.com/ultralytics/ultralytics/blob/4a7ccba0af45834498d7a6362258c46acc993e2c/ultralytics/utils/loss.py)

**Comment by glenn-jocher** [comment]
@ksv87 hi there! 😊

Indeed, for applying class weights in YOLOv8, you're looking in the right direction by exploring the loss function. Typically, modifications for class weighting would occur in the loss computation.

While YOLOv8's exact implementation details might differ, you can experiment with class weights within the `build_targets` or directly within the loss computation sections of the code. A pseudo-example might look something like this:

```python
# Assume class_weights is a tensor of shape [num_classes] with your predefined weights
class_weights = torch.tensor([1.0, 2.0, 0.5])  # Example weights for 3 classes

# During loss calculation, you might see something like...
loss_cls = F.cross_entropy(prediction, targets, weight=class_weights[targets])
```

Please adjust according to the actual structure and variables in the loss calculation code. This is a simplified example; the actual modification might require a deeper dive into how the predictions and targets are structured in YOLOv8.

Hope this points you in the right direction! Let us know if you have more questions.

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/21082
